packages/scm/src/browser/scm-widget.tsx

```diff
@@ -154,540 +133,45 @@ export class ScmWidget extends ReactWidget implements StatefulWidget {
-                    hover,
-                    args: this.contextMenuArgs,
-                    menu: menus.getMenu(ScmWidget.RESOURCE_GROUP_INLINE_MENU),
-                    commands,
-                    contextKeys,
-                    group
-                }}>
-                    {this.renderChangeCount()}
-                </ScmInlineActions>
-            </div>
-            <div>{group.resources.map(resource => this.renderScmResourceItem(resource))}</div>
-        </div>;
-    }
-
-    protected renderChangeCount(): React.ReactNode {
-        return <div className='notification-count-container scm-change-count'>
-            <span className='notification-count'>{this.props.group.resources.length}</span>
-        </div>;
-    }
-
-    protected renderScmResourceItem(resource: ScmResource): React.ReactNode {
-        const name = this.props.labelProvider.getName(resource.sourceUri);
-        return <ScmResourceComponent
-            key={String(resource.sourceUri)}
-            {...{
-                ...this.props,
-                name,
-                resource
-            }}
-        />;
-    }
 
-    protected readonly contextMenuPath = ScmWidget.RESOURCE_GROUP_CONTEXT_MENU;
-    protected get contextMenuArgs(): any[] {
-        return [this.props.group];
-    }
-}
-
-export class ScmInlineActions extends React.Component<ScmInlineActions.Props> {
-    render(): React.ReactNode {
-        const { hover, menu, args, commands, group, contextKeys, children } = this.props;
-        return <div className='theia-scm-inline-actions-container'>
-            <div className='theia-scm-inline-actions'>
-                {hover && menu.children.map((node, index) => node instanceof ActionMenuNode && <ScmInlineAction key={index} {...{ node, args, commands, group, contextKeys }} />)}
-            </div>
-            {children}
-        </div>;
-    }
-}
-export namespace ScmInlineActions {
-    export interface Props {
-        hover: boolean;
-        menu: CompositeMenuNode;
-        commands: CommandRegistry;
-        group: ScmResourceGroup;
-        contextKeys: ScmContextKeyService;
-        args: any[];
-        children?: React.ReactNode;
-    }
-}
-
-export class ScmInlineAction extends React.Component<ScmInlineAction.Props> {
-    render(): React.ReactNode {
-        const { node, args, commands, group, contextKeys } = this.props;
-        const currentScmResourceGroup = contextKeys.scmResourceGroup.get();
-        contextKeys.scmResourceGroup.set(group.id);
-        try {
-            if (!commands.isVisible(node.action.commandId, ...args) || !contextKeys.match(node.action.when)) {
-                return false;
-            }
-            return <div className='theia-scm-inline-action'>
-                <a className={node.icon} title={node.label} onClick={this.execute} />
-            </div>;
-        } finally {
-            contextKeys.scmResourceGroup.set(currentScmResourceGroup);
+        export interface WidgetDescriptor {
```

Comment:
It should not be necessary. Clients should provider widget factories and implement stateful widget interface, Theia will take case about using `WidgetManager` even for composite widgets.

Reply:
Now removed.